include/sys/onoff.h

```diff
@@ -0,0 +1,611 @@
+ * notification, and instead must periodically check for completion
+ * using onoff_client_fetch_result().
+ *
+ * @param cli pointer to the client state object.
+ */
+static inline void onoff_client_init_spinwait(struct onoff_client *cli)
+{
+	__ASSERT_NO_MSG(cli != NULL);
+
+	*cli = (struct onoff_client){
+		.flags = ONOFF_CLIENT_NOTIFY_SPINWAIT,
+	};
+}
+
+#if defined(CONFIG_POLL) || defined(DOXYGEN)
+/**
+ * @brief Initialize an on-off client to be used for a signal
+ * operation notification.
+ *
+ * Clients that use this initialization will be notified of the
+ * completion of operations submitted through onoff_request() and
+ * onoff_release() through the provided signal.  This capability is
+ * available only when `CONFIG_POLL` is selected.
+ *
+ * @param cli pointer to the client state object.
+ *
+ * @param sigp pointer to the signal to use for notification.  The
+ * value must not be null.  The signal must be reset before the client
+ * object is passed to the on-off service API.
+ */
+static inline void onoff_client_init_signal(struct onoff_client *cli,
+					    struct k_poll_signal *sigp)
+{
+#ifndef DOXYGEN
+	__ASSERT_NO_MSG(cli != NULL);
+	__ASSERT_NO_MSG(sigp != NULL);
+
+	*cli = (struct onoff_client){
+		.async = {
+			.signal = sigp,
+		},
+		.flags = ONOFF_CLIENT_NOTIFY_SIGNAL,
+	};
+#endif /* DOXYGEN */
+}
+#endif /* CONFIG_POLL */
+
+/**
+ * @brief Initialize an on-off client to be used for a callback
+ * operation notification.
+ *
+ * Clients that use this initialization will be notified of the
+ * completion of operations submitted through on-off service API
+ * through the provided callback.  Note that callbacks may be invoked
+ * from various contexts depending on the specific service; see
+ * @ref onoff_client_callback.
+ *
+ * @param cli pointer to the client state object.
+ *
+ * @param handler a function pointer to use for notification.
+ *
+ * @param user_data an opaque pointer passed to the handler to provide
+ * additional context.
+ */
+static inline void onoff_client_init_callback(struct onoff_client *cli,
+					      onoff_client_callback handler,
+					      void *user_data)
+{
+	__ASSERT_NO_MSG(cli != NULL);
+	__ASSERT_NO_MSG(handler != NULL);
+
+	*cli = (struct onoff_client){
+		.async = {
+			.callback = {
+				.handler = handler,
+				.user_data = user_data,
+			},
+		},
+		.flags = ONOFF_CLIENT_NOTIFY_CALLBACK,
+	};
+}
+
+/**
+ * @brief Request a reservation to use an on-off service.
+ *
+ * The return value indicates the success or failure of an attempt to
+ * initiate an operation to request the resource be made available.
+ * If initiation of the operation succeeds the result of the request
+ * operation is provided through the configured client notification
+ * method, possibly before this call returns.
+ *
+ * Note that the call to this function may succeed in a case where the
+ * actual request fails.  Always check the operation completion
+ * result.
+ *
```

Comment:
@mbolivar here's the new attempt to describe the situation.

Reply:
Great example; that really cleared it up for me. Thanks.